### PR TITLE
Fgp 39: Logout

### PR DESCRIPTION
**Pull request by rachelswart**
- Adds JWT to blacklist on logout
- Calls COLA logout
- Removes token from cookies 
- Scheduled job to remove expired JWTs from blacklist 
- Checks if token is blacklisted when checking if it's valid

### CODE REVIEW

File: src/main/java/gov/cabinetofice/gapuserservice/model/Blacklist.java [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+@Getter
+@Setter
+@NoArgsConstructor
+@AllArgsConstructor
+@Builder
+@Entity
+@Table
+public class Blacklist {
+    @Id
+    private String jwt;
```
Comment (GavCookCO) [inline src/main/java/gov/cabinetofice/gapuserservice/model/Blacklist.java:20]: Can we add an integer column to this table and use that as the ID please? JWTs can't be relied on to be unique

File: pom.xml [hunk 2]
```diff
@@ -86,6 +86,16 @@
 			<version>0.11.1</version>
 			<scope>runtime</scope>
 		</dependency>
+		<dependency>
```
Comment (GavCookCO) [inline pom.xml:89]: Had never heard of shedlock before but makes sense if we ever want to load balance the application.

File: src/main/java/gov/cabinetofice/gapuserservice/model/Blacklist.java [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+import java.util.Date;
+
+@EntityListeners(AuditingEntityListener.class)
+@Getter
+@Setter
+@NoArgsConstructor
+@AllArgsConstructor
+@Builder
+@Entity
+@Table
```
Comment (GavCookCO) [inline src/main/java/gov/cabinetofice/gapuserservice/model/Blacklist.java:17]: Can we give this a more descriptive table name such as "token_blacklist" or "jwt_blacklist"

File: src/main/java/gov/cabinetofice/gapuserservice/model/Blacklist.java [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+public class Blacklist {
+    @Id
+    private String jwt;
+
+    @Column
+    private Date expiryDate;
```
Comment (GavCookCO) [inline src/main/java/gov/cabinetofice/gapuserservice/model/Blacklist.java:23]: we should be looking to use the java.time date objects going forward if possible.

File: src/main/java/gov/cabinetofice/gapuserservice/service/JwtBlacklistService.java [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+                .build();
+        this.jwtBlacklistRepository.save(blacklist);
+    }
+
+    public boolean isJwtInBlacklist(final String jwt) {
+        return jwtBlacklistRepository.existsByJwtIs(jwt);
+    }
+
+    public Long deleteExpiredJwts() {
+        return this.jwtBlacklistRepository.deleteByExpiryDateLessThan(new Date(ZonedDateTime.now(clock).toInstant().toEpochMilli()));
```
Comment (GavCookCO) [inline src/main/java/gov/cabinetofice/gapuserservice/service/JwtBlacklistService.java:31]: This doesn't need the `new` keyword, you can make static calls to the `ZonedDateTime` class :)

File: pom.xml [hunk 6]
```diff
@@ -112,7 +122,15 @@
 					<cleanDisabled>false</cleanDisabled>
 				</configuration>
 			</plugin>
-		</plugins>
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-compiler-plugin</artifactId>
+                <configuration>
+                    <source>16</source>
```
Comment (GavCookCO) [inline pom.xml:129]: shouldn't this be 17?

Comment (rachelswart) [reply pom.xml:129]: I'm honestly not sure but I can change it to 17

**Review by GavCookCO — APPROVED** [review]

**Pull request merged**